lib/services.js

```diff
@@ -142,15 +143,26 @@ export default {
      * @param {object} request - request linked to the stream containing
      *                           the data
      * @param {number} size - data size in the stream
+     * @param {object | null } streamingV4Params - if v4 auth, object containing
+     * accessKey, signatureFromRequest, region, scopeDate,
+     * timestamp, and credentialScope
+     * (to be used for streaming v4 auth if applicable)
      * @param {RequestLogger} log - the current request logger
      * @param {function} cb - callback containing result for the next task
      * @return {undefined}
      */
     dataStore(objectMetadata, objectContext,
-              cipherBundle, request, size, log, cb) {
-        assert.strictEqual(arguments.length, 7);
-        data.put(cipherBundle, request, size, objectContext, log, (err,
-                                                     dataRetrievalInfo) => {
+              cipherBundle, request, size, streamingV4Params, log, cb) {
+        assert.strictEqual(arguments.length, 8);
+        let dataStream = request;
+        if (request.headers['x-amz-content-sha256'] ===
+            'STREAMING-AWS4-HMAC-SHA256-PAYLOAD') {
+            const v4Transform = new V4Transform(streamingV4Params, log, cb);
+            request.pipe(v4Transform);
+            dataStream = v4Transform;
```

Comment:
I recommend abstracting `data.put` call into a function and sending with appropriate stream request/v4Transform as param. Otherwise ninja replacing the request stream may bite us back in the future.